plugins/plugin-optimize/lib/css.js

```diff
@@ -0,0 +1,121 @@
+const fs = require('fs');
+const path = require('path');
+const {parse} = require('es-module-lexer');
+const csso = require('csso');
+
+/** Early-exit function that determines, given a set of JS files, if CSS is being imported */
+function hasCSSImport(files) {
+  for (const file of files) {
+    const code = fs.readFileSync(file, 'utf-8');
+    const [imports] = parse(code);
+    for (const {s, e} of imports.filter(({d}) => d === -1)) {
+      const spec = code.substring(s, e);
+      if (spec.endsWith('.css.proxy.js')) return true; // exit as soon as we find one
+    }
+  }
+  return false;
+}
+exports.hasCSSImport = hasCSSImport;
+
+/**
+ * Scans JS for CSS imports, and embeds only what’s needed
+ *
+ * import 'global.css'                       -> (removed; loaded in HTML)
+ * import url from 'global.css'              -> const url = 'global.css'
+ * import {foo, bar} from 'local.module.css' -> const {foo, bar} = 'local.module.css'
+ */
+function transformCSSProxy(file, originalCode) {
```

Comment:
lets walk through this code together, I want to make sure that I understand it and I'm not sure I do at the moment :)